src/NHibernate/Cache/QueryKey.cs

```diff
@@ -11,23 +13,33 @@
 namespace NHibernate.Cache
 {
 	[Serializable]
-	public class QueryKey
+	public class QueryKey : IDeserializationCallback
 	{
 		private readonly ISessionFactoryImplementor _factory;
 		private readonly SqlString _sqlQueryString;
 		private readonly IType[] _types;
 		private readonly object[] _values;
 		private readonly int _firstRow = RowSelection.NoValue;
 		private readonly int _maxRows = RowSelection.NoValue;
+
+		// Sets and dictionaries are populated last during deserialization, causing them to be potentially empty
+		// during the deserialization callback. This causes them to be unreliable when used in hashcode computations.
+		// So better also serialize them as other structures and use those other structures in hashcode computation.
+		// We nonetheless need to also serialize them, because they are externally supplied and may use a custom
+		// comparer, implementation, ...
 		private readonly IDictionary<string, TypedValue> _namedParameters;
 		private readonly ISet<FilterKey> _filters;
+		private IEnumerable<KeyValuePair<string, TypedValue>> _serializedNamedParameters;
+		private IEnumerable<FilterKey> _serializedFilters;
```

Comment:
After deserialization callback is dropped from `FilterKey` this would not be needed.